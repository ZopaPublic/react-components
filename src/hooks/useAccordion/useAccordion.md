### Summary

`useAccordion()` can be used to expand and show hidden content while being accessible.

It can be useful for accordions and in tables with expanding rows. To see it being used along with context see the [`<Accordion>`](/#/Components/Organisms/Accordion).

### Examples

```js
import { Link, Text, useAccordion } from '@zopauk/react-components';

const AccordionExample = () => {
  const { getHeaderProps, getSectionProps, isActiveSection } = useAccordion();

  const items = [
    {
      id: 'one',
      header: 'Header one',
      section: 'example one',
    },
    {
      id: 'two',
      header: 'Header two',
      section: 'example two',
    },
  ];

  return (
    <div aria-label="accordion example">
      {items.map(({ id, header, section, size }, index) => {
        const { ref: headerRef, onClick, ...headerPropsRest } = getHeaderProps(id, index);
        const { ref: sectionRef, ...sectionPropsRest } = getSectionProps(id, index);
        return (
          <div key={id}>
            <Link ref={headerRef} onClick={onClick} {...headerPropsRest}>
              {header}
            </Link>
            <div {...sectionPropsRest}>
              <div ref={sectionRef}>{section}</div>
            </div>
          </div>
        );
      })}
    </div>
  );
};

<AccordionExample />;
```
