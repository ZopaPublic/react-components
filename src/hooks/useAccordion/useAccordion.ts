import { useRef, useState, KeyboardEvent, RefObject } from 'react';
import { isArrowDown, isArrowUp } from '../../helpers/keyboard-keys';
import { mod } from '../../helpers/utils';

export interface AccordionHeaderProps {
  'aria-controls': string;
  'aria-expanded': boolean;
  id: string;
  key: string;
  onClick: () => void;
  onFocus: () => void;
  onKeyDown: (e: KeyboardEvent<HTMLButtonElement>) => void;
  ref: (node: HTMLButtonElement) => void;
}

export type GetAccordionHeaderProps = (id: string, index: number) => AccordionHeaderProps;

interface AccordionSectionStyles {
  overflow: string;
  transition: string;
  height: string;
}

export interface AccordionSectionProps {
  'aria-hidden': boolean;
  'aria-labelledby': string;
  id: string;
  key: string;
  ref: (node: HTMLDivElement) => void;
  role: string;
  style: AccordionSectionStyles;
}

export type GetAccordionSectionProps = (id: string, index: number) => AccordionSectionProps;

export type IsActiveAccordionSection = (index: number) => boolean;

type ActiveSections = number[];

// eslint-disable-next-line
export const useAccordion = (props?: {}) => {
  const headersRefs = useRef<RefObject<HTMLButtonElement>['current'][]>([]).current;

  const getHeaderRef = (index: number) => (node: HTMLButtonElement) => {
    if (node === null) {
      headersRefs.splice(index, 1);
    } else {
      headersRefs[index] = node;
    }
  };

  const sectionsRefs = useRef<RefObject<HTMLDivElement>['current'][]>([]).current;

  const getSectionRef = (index: number) => (node: HTMLDivElement) => {
    if (node === null) {
      sectionsRefs.splice(index, 1);
    } else {
      sectionsRefs[index] = node;
    }
  };

  const [activeSections, updateActiveSections] = useState<ActiveSections>([]);
  const isActiveSection: IsActiveAccordionSection = (index) => activeSections.includes(index);

  const getSectionStyle = (index: number) => {
    const sectionRef = sectionsRefs[index];
    const baseStyle = {
      overflow: 'hidden',
      transition: 'height 200ms linear',
    };
    if (!sectionRef || !isActiveSection(index)) {
      return { ...baseStyle, height: '0px' };
    }
    return { ...baseStyle, height: `${sectionRef.clientHeight}px` };
  };

  const toggleAccordionSection = (index: number) => {
    updateActiveSections((prevSections) =>
      prevSections.includes(index) ? prevSections.filter((i) => i !== index) : prevSections.concat(index),
    );
  };

  const [cursorPosition, updateCursorPosition] = useState<number>(-1);

  const getOnClick = (index: number) => () => {
    updateCursorPosition(index);
    toggleAccordionSection(index);
  };

  const getOnFocus = (index: number) => () => updateCursorPosition(index);

  const getLinkingId = (id: string) => `${id}-section`;

  const focusOnHeader = (nextCursorPosition: number) => {
    const headerRef = headersRefs[nextCursorPosition];
    if (headerRef) {
      headerRef.focus();
    }
  };

  const onKeyDown = (e: KeyboardEvent<HTMLButtonElement>) => {
    const { length } = headersRefs;
    if (isArrowUp(e)) {
      e.preventDefault();
      const nextCursorPosition = mod(cursorPosition - 1, length);
      updateCursorPosition(nextCursorPosition);
      focusOnHeader(nextCursorPosition);
    } else if (isArrowDown(e)) {
      e.preventDefault();
      const nextCursorPosition = mod(cursorPosition + 1, length);
      updateCursorPosition(nextCursorPosition);
      focusOnHeader(nextCursorPosition);
    }
  };

  const getHeaderProps: GetAccordionHeaderProps = (id, index) => ({
    'aria-controls': getLinkingId(id),
    'aria-expanded': isActiveSection(index),
    id,
    key: id,
    onClick: getOnClick(index),
    onFocus: getOnFocus(index),
    onKeyDown,
    ref: getHeaderRef(index),
  });

  const getSectionProps: GetAccordionSectionProps = (id, index) => ({
    'aria-hidden': !isActiveSection(index),
    'aria-labelledby': id,
    id: getLinkingId(id),
    key: getLinkingId(id),
    ref: getSectionRef(index),
    role: 'region',
    style: getSectionStyle(index),
  });

  return {
    getHeaderProps,
    getSectionProps,
    isActiveSection,
  };
};
