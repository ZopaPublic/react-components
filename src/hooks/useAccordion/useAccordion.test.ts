import { renderHook, act } from '@testing-library/react-hooks';

import { useAccordion } from './useAccordion';

describe('useAccordion', () => {
  it('should render correct tab and section props', () => {
    const { result } = renderHook(useAccordion);

    const header = result.current.getHeaderProps('id', 1);
    const section = result.current.getSectionProps('id', 1);

    expect(header).toMatchSnapshot();
    expect(section).toMatchSnapshot();
  });

  it('should open and close espective sections on header clicks', () => {
    const { result } = renderHook(useAccordion);

    const { onClick: firstHeaderClick } = result.current.getHeaderProps('one', 1);

    expect(result.current.isActiveSection(1)).toEqual(false);

    act(() => {
      firstHeaderClick();
    });

    expect(result.current.isActiveSection(1)).toEqual(true);
    const { onClick: secondHeaderClick } = result.current.getHeaderProps('two', 2);
    expect(result.current.isActiveSection(2)).toEqual(false);

    act(() => {
      secondHeaderClick();
    });

    expect(result.current.isActiveSection(2)).toEqual(true);

    act(() => {
      secondHeaderClick();
    });

    expect(result.current.isActiveSection(2)).toEqual(false);
  });

  it('should render matching header and section attributes', () => {
    const { result } = renderHook(useAccordion);

    const section = result.current.getSectionProps('another-id', 3);
    const header = result.current.getHeaderProps('another-id', 3);

    expect(header.id).toEqual(section['aria-labelledby']);
  });
});
