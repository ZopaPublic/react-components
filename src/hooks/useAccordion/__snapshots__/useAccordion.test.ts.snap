// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useAccordion should render correct tab and section props 1`] = `
Object {
  "aria-controls": "id-section",
  "aria-expanded": false,
  "id": "id",
  "key": "id",
  "onClick": [Function],
  "onFocus": [Function],
  "onKeyDown": [Function],
  "ref": [Function],
}
`;

exports[`useAccordion should render correct tab and section props 2`] = `
Object {
  "aria-hidden": true,
  "aria-labelledby": "id",
  "id": "id-section",
  "key": "id-section",
  "ref": [Function],
  "role": "region",
  "style": Object {
    "height": "0px",
    "overflow": "hidden",
    "transition": "height 200ms linear",
  },
}
`;
